auth/src/main/java/com/vimeo/networking2/internal/AuthenticatorImpl.kt

```diff
@@ -230,24 +230,77 @@ internal class AuthenticatorImpl(
 
             call.enqueueError(apiError, callback)
         } else {
-            call.enqueue(callback)
+            call.enqueueWithAccountStore(callback)
         }
     }
 
     override fun authenticateWithCodeGrant(uri: String, callback: VimeoCallback<VimeoAccount>): VimeoRequest {
         val httpUrl: HttpUrl = HttpUrl.parse(uri) ?: return localVimeoCallAdapter.enqueueEmptyUri(callback)
-        val authorizationCode = httpUrl.queryParameter("code")
+        val authorizationCode = httpUrl.queryParameter(PARAM_CODE)
             ?: return localVimeoCallAdapter.enqueueEmptyUri(callback)
         return authService.authenticateWithCodeGrant(
             basicAuthHeader,
             redirectUri,
             authorizationCode,
             GrantType.AUTHORIZATION_CODE
-        ).enqueue(AccountStoringVimeoCallback(accountStore, callback))
+        ).enqueueWithAccountStore(callback)
     }
 
     override fun createCodeGrantAuthorizationUri(responseCode: String): String {
-        return authService.codeGrantRequest(clientId, redirectUri, responseCode, scopes).request().url().toString()
+        return authService.codeGrantRequest(clientId, redirectUri, responseCode, scopes).url.toString()
+    }
+
+    override fun fetchSsoDomain(domain: String, callback: VimeoCallback<SsoDomain>): VimeoRequest {
+        val invalidAuthParams = mapOf(
+            AuthParam.DOMAIN to domain
+        ).validate()
+
+        val call = authService.getSsoDomain(basicAuthHeader, domain)
+
+        return if (invalidAuthParams.isNotEmpty()) {
+            val apiError = ApiError(
+                developerMessage = "SSO domain fetch error",
+                invalidParameters = invalidAuthParams
+            )
+            call.enqueueError(apiError, callback)
+        } else {
+            call.enqueue(callback)
+        }
+    }
+
+    override fun createSsoGrantAuthorizationUri(ssoDomain: SsoDomain, responseCode: String): String {
+        val connectUrl = requireNotNull(ssoDomain.connectUrl) { "SsoDomain.connectUrl cannot be null" }
+        return authService.ssoGrantRequest(connectUrl, redirectUri, responseCode).url.toString()
+    }
+
+    override fun ssoCodeGrant(
+        uri: String,
+        marketingOptIn: Boolean,
+        callback: VimeoCallback<VimeoAccount>
+    ): VimeoRequest {
+        val httpUrl: HttpUrl = HttpUrl.parse(uri) ?: return localVimeoCallAdapter.enqueueEmptyUri(callback)
+        val authorizationCode = httpUrl.queryParameter(PARAM_CODE).orEmpty()
+        val invalidAuthParams = mapOf(
+            AuthParam.AUTHORIZATION_CODE to authorizationCode,
+            AuthParam.REDIRECT_URI to redirectUri
+        ).validate()
```

Comment:
This original approach to local validation feels sloppy and hard to maintain. I had to add 3 entries to the `AuthParam` enum for this PR and I am not sure it's worth the effort. It complicates the implementation of this class, increasing the burden on the reader and the maintenance effort. 

The simpler approach I took for URIs in `VimeoApiClient` (used for the httpUrl here) is easier to maintain and read.
```
val httpUrl: HttpUrl = HttpUrl.parse(uri) ?: return localVimeoCallAdapter.enqueueEmptyUri(callback)
```
I'm wondering if perhaps we should generify it to take a parameter name like
```
val httpUrl: HttpUrl = HttpUrl.parse(uri) ?: return localVimeoCallAdapter.enqueueInvalidParameter("uri", callback)
```
I'm not really sure how to report the empty authorization code in this case though as it's a parameter in the `uri` parameter. So would we add a secondary message to the error or would we just call out the `authorizationCode` as being missing?

Reply:
The issue I see with a hybrid approach like this is that it cannot guarantee non nullity. So for example, in `VimeoApiClient.editAlbum`, currently we have 
```kotlin
val uri = album.uri.notEmpty() ?: return localVimeoCallAdapter.enqueueEmptyUri(callback)
```
which verifies that the URI is neither empty nor null. If we were to go with this hybrid approach or the current approach used in the `Authenticator`, then we would need an additional `album.uri!!` or `requireNotNull(album.uri)` when passing it to the service in order to satisfy the type system. Needing to use a lot of these assertions usually is an indication that there is an issue with the design. `VimeoApiClient` would need 16 of these assertions, out of a total of 68 error checks. Another thought, if we wanted to move away from throwing an exception if a user provides an `UNKNOWN` enum value (see `StringValue`) to a function that supplies enums, we will also need to support some sort of null assertion.

Do you have any thoughts on how to proceed based on that @howardrigberg?

Reply:
Good points. I think your concern about the non-null guarantee on a `String` could be addressed in the design of the `ApiParameter` class as follows:

```
data class ApiParameter(
    val fieldName: String,
    val fieldValue: String?,
    val errorCodeType: ErrorCodeType = ErrorCodeType.DEFAULT,
    var validatedFieldValue: String = ""
) 
```
It is a little odd at first glance, but when used in the context of the validation check from `validateParameters`, you'd be guaranteed to get what you want on the other end if you bail out whenever you get a non-null `VimeoRequest` back from `validateParameters`. The `validateParameters` method would set `isValidated` for each `ApiParameter` as it is checked.
```
fun <T> LocalVimeoCallAdapter.validateParameters(
        callback: VimeoCallback<T>,
        developerMessage: String?,
        vararg params: ApiParameter
    ): VimeoRequest? {
        val invalidParameters = mutableListOf<InvalidParameter>()
        for (param in params) {
            if (param.fieldValue.isNullOrBlank()) {
                invalidParameters.add(InvalidParameter(
                    field = param.fieldName,
                    errorCode = param.errorCodeType.value,
                    developerMessage = "${param.fieldName} is empty or null"
                ))
            } else {
               param.validatedFieldValue = param.fieldValue
           }
        }
        return if (invalidParameters.isNotEmpty()) {
            enqueueError(ApiError(
                invalidParameters = invalidParameters,
                developerMessage = developerMessage
            ), callback)
        } else {
            null
        }
    }
```

The revised way of interacting would be a little more cumbersome because you'd be keeping references around to the ApiParameters rather than creating them inline to pass into `validateParameters`:
```
val uriParam = ApiParameter("uri", httpUrl?.toString(), ErrorCodeType.INVALID_URI)
val authCodeParam = ApiParameter("authorization_code", authorizationCode)
val redirectUriParam = ApiParameter("redirect_uri", redirectUri)
localVimeoCallAdapter.validateParameters(callback, "SSO code grant error", uriParam, authCodeParam, redirectUriParam)?.let { return it }

iExpectANonNullStringParam(uriParam.validatedFieldValue)
```